File: SWEA/D4/1868. 파핑파핑 지뢰찾기/파핑파핑 지뢰찾기.py
import queue

dx = [-1, -1, -1, 0, 0, 1, 1, 1]
dy = [-1, 0, 1, -1, 1, -1, 0, 1]

T = int(input())


def checkBomb(n, arr):
    for i in range(n):
        for j in range(n):
            if arr[i][j] == ".":
                cnt = 0
                for h in range(8):
                    nx = i + dx[h]
                    ny = j + dy[h]
                    if nx >= 0 and ny >= 0 and nx < n and ny < n and arr[nx][ny] == "*":
                        cnt += 1
                arr[i][j] = cnt


def bfs(i, j, visited):
    q = queue.Queue()
    q.put([i, j])
    visited[i][j] = True

    while q.empty() == False:
        x, y = q.get()
        for h in range(8):
            nx = x + dx[h]
            ny = y + dy[h]
            if (
                nx >= 0
                and ny >= 0
                and nx < n
                and ny < n
                and arr[nx][ny] != "*"
                and visited[nx][ny] == False
            ):
                visited[nx][ny] = True
                if arr[nx][ny] == 0:
                    q.put([nx, ny])


for test_case in range(1, T + 1):
    n = int(input())
    arr = [list(input()) for _ in range(n)]
    visited = [[False] * n for _ in range(n)]
    result = 0

    checkBomb(n, arr)

    for i in range(n):
        for j in range(n):
            if arr[i][j] == 0 and visited[i][j] == False:
                bfs(i, j, visited)
                result += 1

    for i in range(n):
        for j in range(n):
            if visited[i][j] == False and arr[i][j] != "*":
                result += 1

    print("#{} {}".format(test_case, result))

File: SWEA/D4/1868. 파핑파핑 지뢰찾기/파핑파핑 지뢰찾기.java
import java.util.Scanner;

class Solution {
    static int n;
    static char[][] arr;
    static boolean[][] visited;
    static int[] dx = {-1,-1,-1,0,0,1,1,1};
    static int[] dy = {-1,0,1,-1,1,-1,0,1};

    public static void main(String args[]) throws Exception {
        Scanner sc = new Scanner(System.in);
        int T;
        T=sc.nextInt();

        for(int test_case = 1; test_case <= T; test_case++)
        {
            int result = 0;
            n = sc.nextInt();
            arr = new char[n][n];
            visited = new boolean[n][n];
            for(int i = 0; i<n; i++) {
                String s = sc.next();
                for(int j = 0; j<n; j++) {
                    arr[i][j] = s.charAt(j);
                }
            }

            checkBomb();

            for(int i = 0; i<n; i++){
                for(int j = 0; j<n; j++){
                    if(arr[i][j] == '0' && !visited[i][j]) {
                        dfs(i,j);
                        result++;
                    }
                }
            }

            for(int i = 0; i<n; i++){
                for(int j = 0; j<n; j++){
                    if(arr[i][j] != '0' && arr[i][j] != '*' && !visited[i][j]) {
                        result++;
                    }
                }
            }

            System.out.println("#" + test_case + " " + result);
        }
    }

    public static void checkBomb() {
        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++) {
                if(arr[i][j] == '.') {
                    int cnt = 0;
                    for(int h = 0; h<8; h++) {
                        int nx = i + dx[h];
                        int ny = j + dy[h];
                        if(nx >= 0 && ny >= 0 && nx < n && ny < n && arr[nx][ny] == '*')
                            cnt++;
                    }
                    arr[i][j] = (char)(cnt + '0');
                }
            }
        }
    }

    static void dfs(int i, int j) {
        visited[i][j] = true;

        if(arr[i][j] == '0') {
            for(int h = 0; h < 8; h++) {
                int nx = i + dx[h];
                int ny = j + dy[h];
                if(nx >= 0 && ny >= 0 && nx < n && ny < n && arr[nx][ny] != '*' && !visited[nx][ny]) {
                    dfs(nx,ny);
                }
            }
        }
    }
}